<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Inga Patarcic" />

<meta name="date" content="2017-11-28" />

<title>reg2gene</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">reg2gene</h1>
<h4 class="author"><em>Inga Patarcic</em></h4>
<h4 class="date"><em>2017-11-28</em></h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This document describes <strong>reg2gene</strong> - an R package which can be used to predict the target genes for regulatory elements genome-wide.</p>
<p>To link enhancer activity and gene expression it uses <strong>correlation</strong> and/or other statistical modelling approaches ( <strong>elastic net</strong> and <strong>random forests</strong>) of <strong>enhancer activity ~ gene expression</strong> across different cell types. This approach is based on the observation that enhancers show very high tissue specificity and the level of their activity correlates with gene expression (Visel et al. 2009., Ernst et al. 2011).</p>
<p>All analyses can be done easily for different types of enhancer-related chromatin marks ( <strong>H3K4me1</strong>, <strong>H3K27ac</strong>, <strong>DNA methylation</strong>, <strong>DHS</strong>, etc.) and for different sources of data (Roadmap, Blueprint or in-house datasets). Single-model information (gene-enhancer pairs) can be aggregated by means of <strong>meta-analysis</strong> across different data-sources or by <strong>majority voting</strong> decisions across both different modelling procedures or different data-types, e.g.chromatin marks. To benchmark reported associations, external benchmark dataset can be used to identify overlapping pairs in the genome (predicted gene~enhancer pairs that overlap in location with benchmark pair), whereas <strong>confusion matrix</strong> can be calculated to assess performance of the modelling procedure based on the used external benchmark dataset. Often, information about the 3D genome architecture, generated by chromosome conformation capture and related techniques, can be used as a proxy of enhancer-mediated regulation of gene expression (Dekker et al. 2002, Dosie et al. 2006, Simonis et al. 2006, Fullwood et al. 2009, Lieberman-Aiden et al. 2009), and thus can represent a good benchmark dataset <strong>to benchmark results</strong> of reg2gene modelling approach.</p>
<p>To achieve aforementioned functionality, this package gathers functions that:</p>
<pre><code>* a) quantify gene expression and enhancer activity based on input 
bigwig files: regActivity() , bwToGeneExp() ,regActivityAroundTSS(),

* b) model and test quantified gene expression ~ enhancer activity 
associations: associateReg2Gene(),

* c) combine different models in terms of meta-association or majority
voting decisions: metaAssociations(), voteAssociations()

* d) benchmark reported associations with predefined benchmark dataset:
benchmarkGI()

* e) assess modelling performance via calculating and reporting confusion 
matrix for tested benchmarked dataset: confusionMatrix(). </code></pre>
</div>
<div id="add-projects-flowchart" class="section level1">
<h1>Add projects flowchart</h1>
</div>
<div id="data-input" class="section level1">
<h1>Data input</h1>
<div id="input-data-type-1-granges-objects" class="section level2">
<h2>INPUT DATA TYPE 1: GRanges objects</h2>
<p>This package comes with 2 toy datasets, GRanges objects:</p>
<p>GRReg1_toy</p>
<pre><code>## GRanges object with 9 ranges and 7 metadata columns:
##       seqnames     ranges strand |          reg       name2        name
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;GRanges&gt; &lt;character&gt; &lt;character&gt;
##   [1]     chr1 [  1,   4]      * |     chr1:5-9   TEST_Reg1   TEST_Reg1
##   [2]     chr1 [  1,   4]      * |   chr1:30-31   TEST_Reg2   TEST_Reg2
##   [3]     chr1 [ 15,  20]      * |   chr1:21-25   TEST_Reg3   TEST_Reg3
##   [4]     chr1 [ 24,  29]      * |    chr1:5-10   TEST_Reg4   TEST_Reg4
##   [5]     chr1 [ 30,  31]      * |   chr1:31-32   TEST_Reg5   TEST_Reg5
##   [6]     chr1 [  5,   9]      * |   chr1:31-32   TEST_Reg6   TEST_Reg6
##   [7]     chr1 [100, 101]      * | chr1:102-103   TEST_Reg7   TEST_Reg7
##   [8]     chr1 [200, 201]      * | chr1:202-203   TEST_Reg8   TEST_Reg8
##   [9]     chr2 [  1,   4]      * |     chr2:5-9   TEST_Reg9   TEST_Reg9
##       Bench1Exp Filter2Exp Bench2Exp Filter1Exp
##       &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;
##   [1]         2          1         1          1
##   [2]         1          1         1          1
##   [3]         0          1         1          0
##   [4]         0          1         1          1
##   [5]         1          1         1          1
##   [6]         2          1         1          1
##   [7]         3          1         1          1
##   [8]         2          1         1          1
##   [9]         2          1         1          1
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>and GRReg2_toy</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reg2gene<span class="op">::</span>GRReg2_toy</code></pre></div>
<pre><code>## GRanges object with 14 ranges and 3 metadata columns:
##        seqnames    ranges strand |        reg       name2        name
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;GRanges&gt; &lt;character&gt; &lt;character&gt;
##    [1]     chr1  [ 1,  3]      * |  chr1:5-10   TEST_Reg1   TEST_Reg1
##    [2]     chr1  [ 5,  9]      * |   chr1:1-2   TEST_Reg2   TEST_Reg2
##    [3]     chr1  [26, 30]      * | chr1:31-32   TEST_Reg3   TEST_Reg3
##    [4]     chr1  [ 1,  3]      * | chr1:31-32   TEST_Reg4   TEST_Reg4
##    [5]     chr1  [31, 32]      * |   chr1:1-3   TEST_Reg5   TEST_Reg5
##    ...      ...       ...    ... .        ...         ...         ...
##   [10]     chr2  [26, 30]      * | chr2:31-32  TEST_Reg10  TEST_Reg10
##   [11]     chr2  [ 1,  3]      * | chr2:31-32  TEST_Reg11  TEST_Reg11
##   [12]     chr2  [31, 32]      * |   chr2:1-3  TEST_Reg12  TEST_Reg12
##   [13]     chr2  [ 5, 10]      * | chr2:31-32  TEST_Reg13  TEST_Reg13
##   [14]     chr2  [31, 32]      * |  chr2:5-10  TEST_Reg14  TEST_Reg14
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>with genome locations preselected in such way that all possible outputs ofthe benchmarking procedure is captured.</p>
<p>*Colums which correspond to the expected outcome of the benchmarking procedure are:</p>
<p>Bench1Exp,</p>
<p>Filter2Exp,</p>
<p>Bench2Exp,</p>
<p>Filter1Exp</p>
<p>Where Bench1 andFilter1 corresponds to the benchmarking of GRReg1_toy with GRReg2_toy, whereas Bench2 and Filter2 corresponds to the benchmarking of GRReg1_toy with itself.</p>
<p>For example,following object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reg2gene<span class="op">::</span>GRReg1_toy[<span class="dv">7</span>]</code></pre></div>
<pre><code>## GRanges object with 1 range and 7 metadata columns:
##       seqnames     ranges strand |          reg       name2        name
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;GRanges&gt; &lt;character&gt; &lt;character&gt;
##   [1]     chr1 [100, 101]      * | chr1:102-103   TEST_Reg7   TEST_Reg7
##       Bench1Exp Filter2Exp Bench2Exp Filter1Exp
##       &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;
##   [1]         3          1         1          1
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>is benchmarked 3 times with GRReg2_toy</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reg2gene<span class="op">::</span>GRReg2_toy[<span class="dv">10</span><span class="op">:</span><span class="dv">12</span>]</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames    ranges strand |        reg       name2        name
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;GRanges&gt; &lt;character&gt; &lt;character&gt;
##   [1]     chr2  [26, 30]      * | chr2:31-32  TEST_Reg10  TEST_Reg10
##   [2]     chr2  [ 1,  3]      * | chr2:31-32  TEST_Reg11  TEST_Reg11
##   [3]     chr2  [31, 32]      * |   chr2:1-3  TEST_Reg12  TEST_Reg12
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="input-data-type-2-.bigwig-files-as-an-input" class="section level2">
<h2>INPUT DATA TYPE 2: .bigWig files as an input</h2>
<p>A second necesarry argument for quantification functions is a named list of BigWig file paths is used as an input to link bigwig locations.</p>
<p>As example shows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(reg2gene)

test.bw &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;test.bw&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;reg2gene&quot;</span>)
test2.bw &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;test2.bw&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;reg2gene&quot;</span>)

regActivityInputExample &lt;-<span class="st"> </span><span class="kw">c</span>(test.bw,test2.bw)  

<span class="kw">print</span>(regActivityInputExample)</code></pre></div>
<pre><code>## [1] &quot;/home/ipatarc/R/x86_64-redhat-linux-gnu-library/3.4/reg2gene/extdata/test.bw&quot; 
## [2] &quot;/home/ipatarc/R/x86_64-redhat-linux-gnu-library/3.4/reg2gene/extdata/test2.bw&quot;</code></pre>
</div>
<div id="input-data-type-3-ginteractions-objects" class="section level2">
<h2>INPUT DATA TYPE 3: GInteractions objects</h2>
<p>Input data for quantification functions are GRanges objects.</p>
<p>However, modelling procedure outputs, by default, GInteractions object from the InteractionSet package.</p>
<div id="introducing-a-ginteractions-object" class="section level3">
<h3>Introducing a GInteractions object</h3>
<p>Toy or any other GRanges objects can be easily organized as GInteractions objects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw">require</span>(InteractionSet)

    GRReg1_toyGI &lt;-<span class="st"> </span>reg2gene<span class="op">::</span>GRReg1_toy
    
    GRReg1_toyGI &lt;-<span class="st"> </span><span class="kw">GInteractions</span>(GRReg1_toyGI,GRReg1_toyGI<span class="op">$</span>reg)

  GRReg1_toyGI[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>## GInteractions object with 3 interactions and 7 metadata columns:
##       seqnames1   ranges1     seqnames2   ranges2 | anchor1.reg
##           &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt; |   &lt;GRanges&gt;
##   [1]      chr1  [ 1,  4] ---      chr1  [ 5,  9] |    chr1:5-9
##   [2]      chr1  [ 1,  4] ---      chr1  [30, 31] |  chr1:30-31
##   [3]      chr1  [15, 20] ---      chr1  [21, 25] |  chr1:21-25
##       anchor1.name2 anchor1.name anchor1.Bench1Exp anchor1.Filter2Exp
##         &lt;character&gt;  &lt;character&gt;         &lt;numeric&gt;          &lt;numeric&gt;
##   [1]     TEST_Reg1    TEST_Reg1                 2                  1
##   [2]     TEST_Reg2    TEST_Reg2                 1                  1
##   [3]     TEST_Reg3    TEST_Reg3                 0                  1
##       anchor1.Bench2Exp anchor1.Filter1Exp
##               &lt;numeric&gt;          &lt;numeric&gt;
##   [1]                 1                  1
##   [2]                 1                  1
##   [3]                 1                  0
##   -------
##   regions: 14 ranges and 0 metadata columns
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</div>
<div id="quantification-and-data-integration" class="section level1">
<h1>1. Quantification and data integration</h1>
<p><strong>QUANTIFICATION:</strong> Before running a modelling analysis, for genes and enhancers of interest, both, geneexpression and enhancer activity need to be quatified (and normalized).</p>
<p><strong>Data integration:</strong> Afterwards, all enhancers and their activities need to be stored in the GRanges object together with gene of interest and its calculated expression levels (look below at 1c)</p>
<div id="a.-quantifying-gene-expression-using-bwtogeneexp" class="section level3">
<h3>1a. Quantifying gene expression using bwToGeneExp()</h3>
<p>To quantify gene expression a bwToGeneExp() can be used. In short, this function firstly quantifies <strong>exon expressions</strong> over pre-defined exon regions using signal from RNA-Seq tracks (bigwig files), then it sums over all exons of a gene of interest to obtain levels of <strong>gene expression</strong>. With this function, gene expression levels can be quantified over a set of samples, cell types or conditions (list of .bigWig files).</p>
<p><strong>IN DETAIL:</strong> For a gene of interest and across all input samples (.bw files) individaully,the expression is firstly calculated for all exon regions definded in the input GRanges object (exons argument, details below). For quantificatons of enhancer activity over this exon regions, a ScoreMatrixBin() from the genomation package is used.</p>
<p>This is followed by per gene expression quantification as follows:</p>
<p><span class="math inline">\(Gene Expression=\sum_{i=1}^n (Exon Expression_n * ExonLength_n)/(GeneLength)\)</span></p>
<p><span class="math inline">\(GeneLength=\sum_{i=1}^n ExonLength_n\)</span></p>
<p>e.g. mean exon expression scores are multiplied by the exon length,summed together and divided by the gene length(a sum of all exon lengths for that gene) After quantifying gene expression, a normalization step (“quantile” or “ratio”) can be runned.</p>
<p>NOTES: This function might be extended to work with BAM files in the future, however now it works only with relevant .bigWig files RNA-Seq .bw files can originate from stranded,unstranded or mixed libraries.</p>
<p><strong>INPUT DATA DETAILS (min arguments):</strong></p>
<div id="exons" class="section level4">
<h4><strong>1. exons</strong></h4>
<p>Information about exon and gene coordinates should be a GInteractions object from the InteractionSet package, with Anchor1 representing the exon locations, and Anchor2 represents location of the corresponding gene.</p>
<pre><code>## GInteractions object with 1 interaction and 3 metadata columns:
##       seqnames1   ranges1     seqnames2   ranges2 | anchor1.reg
##           &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt; |   &lt;GRanges&gt;
##   [1]      chr1    [1, 4] ---      chr1    [5, 9] |    chr1:5-9
##       anchor1.name2 anchor1.name
##         &lt;character&gt;  &lt;character&gt;
##   [1]     TEST_Reg1    TEST_Reg1
##   -------
##   regions: 2 ranges and 0 metadata columns
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Optionally, the input GRanges object that contains exon regions over which the expression will be calculated needs to contain a meta-data columns with the following information:</p>
<pre><code>1) reg - GRanges object - corresponding gene location, or TSS location.
2) name (character)- ENSEMBL or other gene identifier; 
3) name2(character,optional) - 2nd gene identifier, </code></pre>
<p>Example:</p>
<pre><code>## GRanges object with 1 range and 3 metadata columns:
##       seqnames    ranges strand |       reg       name2        name
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;GRanges&gt; &lt;character&gt; &lt;character&gt;
##   [1]     chr1    [1, 4]      * |  chr1:5-9   TEST_Reg1   TEST_Reg1
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>It is strongly suggested to adjust seqlengths of this object to be equal to the seqlenghts <em>Hsapiens</em> from the appropriate package (BSgenome.Hsapiens.UCSC.hg19 or whatever needed version).</p>
</div>
<div id="geneactsignals" class="section level4">
<h4><strong>2. geneActSignals</strong></h4>
<p>A second necesarry argument for quantification functions is a named list of BigWig file paths is used as an input to link bigwig locations.</p>
<p>As example shows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(reg2gene)

test.bw &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;test.bw&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;reg2gene&quot;</span>)
test2.bw &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;test2.bw&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;reg2gene&quot;</span>)

regActivityInputExample &lt;-<span class="st"> </span><span class="kw">c</span>(test.bw,test2.bw)  

<span class="kw">print</span>(regActivityInputExample)</code></pre></div>
<pre><code>## [1] &quot;/home/ipatarc/R/x86_64-redhat-linux-gnu-library/3.4/reg2gene/extdata/test.bw&quot; 
## [2] &quot;/home/ipatarc/R/x86_64-redhat-linux-gnu-library/3.4/reg2gene/extdata/test2.bw&quot;</code></pre>
</div>
</div>
<div id="output" class="section level3">
<h3><strong>OUTPUT:</strong></h3>
<p>When quantification using bwToGeneExp() is performed the previous toy examples of input GRanges objects (regTSS_toy) and a list of bigWig files (regActivityInputExample) a following result is obtained:</p>
<pre><code>## GRanges object with 3 ranges and 4 metadata columns:
##       seqnames    ranges strand |        name       name2      test
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
##   [1]     chr1    [0, 1]      + |   TEST_Reg1   TEST_Reg1 0.3333333
##   [2]     chr2    [8, 9]      + |   TEST_Reg3   TEST_Reg3 4.8333333
##   [3]     chr1    [4, 5]      - |   TEST_Reg5   TEST_Reg5 1.5000000
##           test2
##       &lt;numeric&gt;
##   [1]  2.666667
##   [2] 11.666667
##   [3]  5.000000
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>For each gene, information is quantified first across exons and then summed per gene to quantify gene expression levels.</p>
<p>In the following toy example there are 3 genes: TEST_Reg1, TEST_Reg3 and TEST_Reg5 with different number of exons:</p>
<p>TEST_Reg1 has 2 exons,</p>
<p>TEST_Reg3 has 1 exons,</p>
<p>TEST_Reg3 has 5 exons,</p>
<p>Notice that output GRanges object now has 3 genomic ranges reported which correspond to the toy genes analyzed: TEST_Reg1,TEST_Reg3 and TEST_Reg5. For these genes TSS is location is reported in the genomic ranges object.</p>
<p>Additionally, if exons input is of GInteractions class object, the same output GRanges object is obtained:</p>
<pre><code>## GRanges object with 3 ranges and 4 metadata columns:
##       seqnames    ranges strand |        name       name2 CellType1
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
##   [1]     chr1    [0, 1]      + |   TEST_Reg1   TEST_Reg1 0.3333333
##   [2]     chr2    [8, 9]      + |   TEST_Reg3   TEST_Reg3 4.8333333
##   [3]     chr1    [4, 5]      - |   TEST_Reg5   TEST_Reg5 1.5000000
##       CellType2
##       &lt;numeric&gt;
##   [1]  2.666667
##   [2] 11.666667
##   [3]  5.000000
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<pre><code>## GInteractions object with 6 interactions and 5 metadata columns:
##       seqnames1   ranges1     seqnames2   ranges2 | anchor1.reg
##           &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt; |   &lt;GRanges&gt;
##   [1]      chr1  [ 1,  4] ---      chr1   [1,  4] |  chr1:1-4:+
##   [2]      chr1  [ 7,  8] ---      chr1   [1,  4] |  chr1:1-4:+
##   [3]      chr2  [ 9, 14] ---      chr2   [9, 14] | chr2:9-14:+
##   [4]      chr1  [15, 20] ---      chr1   [1,  4] |  chr1:1-4:-
##   [5]      chr1  [ 1,  4] ---      chr1   [1,  4] |  chr1:1-4:-
##   [6]      chr1  [15, 20] ---      chr1   [1,  4] |  chr1:1-4:-
##       anchor1.name anchor1.name2        name       name2
##        &lt;character&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt;
##   [1]    TEST_Reg1     TEST_Reg1   TEST_Reg1   TEST_Reg1
##   [2]    TEST_Reg1     TEST_Reg1   TEST_Reg1   TEST_Reg1
##   [3]    TEST_Reg3     TEST_Reg3   TEST_Reg3   TEST_Reg3
##   [4]    TEST_Reg5     TEST_Reg5   TEST_Reg5   TEST_Reg5
##   [5]    TEST_Reg5     TEST_Reg5   TEST_Reg5   TEST_Reg5
##   [6]    TEST_Reg5     TEST_Reg5   TEST_Reg5   TEST_Reg5
##   -------
##   regions: 5 ranges and 0 metadata columns
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<pre><code>## GRanges object with 3 ranges and 4 metadata columns:
##       seqnames    ranges strand |        name       name2      test
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
##   [1]     chr1    [0, 1]      + |   TEST_Reg1   TEST_Reg1 0.3333333
##   [2]     chr2    [8, 9]      + |   TEST_Reg3   TEST_Reg3 4.8333333
##   [3]     chr1    [4, 5]      - |   TEST_Reg5   TEST_Reg5 1.5000000
##           test2
##       &lt;numeric&gt;
##   [1]  2.666667
##   [2] 11.666667
##   [3]  5.000000
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p><strong>INPUT DATA DETAILS (additional arguments):</strong></p>
<div id="sampleids-argument" class="section level4">
<h4>sampleIDs argument</h4>
<p>Additional argumets in bwToGeneExp() adjust for names of RNA-Seq libraries (argunment <strong>sampleIDs</strong>); whether stranded or/and unstranded libraries are included, normalization procedure and parallel processing.</p>
<p>For example, <strong>sampleIDs</strong> by defult uses basenames of the paths for input .bigWig files as a unique sample ids/names. However, a vector of unique sample ids/names(.bw files) can be used as well. It is necessary that values in this vector are ordered as the .bigWig file paths are. The previous toy example unique sample ids:test and test2 can be modified with the sampleIDs argument to <strong>CellType</strong> and <strong>CellType2</strong> as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test.bw &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;test.bw&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;reg2gene&quot;</span>)
test2.bw &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;test2.bw&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;reg2gene&quot;</span>)
regTSS_toy &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;chr1&quot;</span>,<span class="dv">2</span>),<span class="st">&quot;chr2&quot;</span>,<span class="kw">rep</span>(<span class="st">&quot;chr1&quot;</span>,<span class="dv">3</span>)),
                      <span class="kw">IRanges</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">15</span>,<span class="dv">1</span>,<span class="dv">15</span>),<span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">8</span>,<span class="dv">14</span>,<span class="dv">20</span>,<span class="dv">4</span>,<span class="dv">20</span>)),
                                            <span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;+&quot;</span>,<span class="dv">3</span>),<span class="kw">rep</span>(<span class="st">&quot;-&quot;</span>,<span class="dv">3</span>)))
regTSS_toy<span class="op">$</span>reg &lt;-<span class="st">  </span>regTSS_toy[<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">5</span>)]
regTSS_toy<span class="op">$</span>name2 &lt;-<span class="st"> </span>regTSS_toy<span class="op">$</span>name &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;TEST_Reg&quot;</span>,
                                        <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">5</span>))
                                        
                                        
 <span class="kw">bwToGeneExp</span>(<span class="dt">exons =</span> regTSS_toy,<span class="dt">geneActSignals =</span> <span class="kw">c</span>(test.bw,test2.bw),
       <span class="dt">sampleIDs=</span><span class="kw">c</span>(<span class="st">&quot;CellType1&quot;</span>,<span class="st">&quot;CellType2&quot;</span>))</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 4 metadata columns:
##       seqnames    ranges strand |        name       name2 CellType1
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
##   [1]     chr1    [0, 1]      + |   TEST_Reg1   TEST_Reg1 0.3333333
##   [2]     chr2    [8, 9]      + |   TEST_Reg3   TEST_Reg3 4.8333333
##   [3]     chr1    [4, 5]      - |   TEST_Reg5   TEST_Reg5 1.5000000
##       CellType2
##       &lt;numeric&gt;
##   [1]  2.666667
##   [2] 11.666667
##   [3]  5.000000
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="libstrand-argument" class="section level4">
<h4>libStrand argument</h4>
<p>RNA-Seq stranded and unstranded libraries allowed. BUT!!! It is crucial that forward and reverse RNA-Seq libraries are listed in one on top of other. For example,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sampleIDs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;/Reverse1.bw&quot;</span>,<span class="st">&quot;/Forward1.bw&quot;</span>,<span class="st">&quot;/Reverse2.bw&quot;</span>,<span class="st">&quot;/Forward2.bw&quot;</span>,
               <span class="st">&quot;Unstranded1&quot;</span>)

sampleIDs</code></pre></div>
<pre><code>## [1] &quot;/Reverse1.bw&quot; &quot;/Forward1.bw&quot; &quot;/Reverse2.bw&quot; &quot;/Forward2.bw&quot;
## [5] &quot;Unstranded1&quot;</code></pre>
<p>needs to be accompanied with</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">libStrand &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>,<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>,<span class="st">&quot;*&quot;</span>)

libStrand</code></pre></div>
<pre><code>## [1] &quot;+&quot; &quot;-&quot; &quot;+&quot; &quot;-&quot; &quot;*&quot;</code></pre>
<div id="othre-arguments-for-bwtogeneexpnormalize-summaryoperation-and-mc.cores" class="section level5">
<h5>Othre arguments for bwToGeneExp():normalize, summaryOperation and mc.cores</h5>
</div>
</div>
<div id="normalization-procedure" class="section level4">
<h4>Normalization procedure:</h4>
<p>Normalization procedure in not performed by default but 2 normalizations can be optionally used: <strong>“quantile”</strong> and <strong>“ratio”</strong>. If normalize argument is set to be “quantile” activity measures are <strong>quantile normalized</strong> as implemented in the normalize.quantiles() from preprocessCore package. If normalize argument is set to be“ratio” then <strong>“median ratio method”</strong> implemented as estimateSizeFactorsForMatrix from DESeq2 package is used to normalize activity measures. <strong>NOTE</strong>: Normalization is runned on the level of gene expression.</p>
</div>
<div id="summaryoperation" class="section level4">
<h4>summaryOperation:</h4>
<p>This argument designates which summary operation should be used over the regions Currently, only “mean” is available, but “median” or “sum” will be implemented in the future.</p>
</div>
</div>
<div id="b.-quantifying-enhancer-activity" class="section level3">
<h3>1b. Quantifying enhancer activity</h3>
<p>On the other hand, for pre-defined GRanges object which contains info about enhancer regions in the genome enhancer activity can be quantified based on the bigWig tracks of ChIP-Seq signals for histone modifications (such as H3K27ac and H3K4me1, but for any other histone modification as well), of DNase-seq signals and of bisulfite-sequencing results for DNA methylation. To quantify enhancer activity regActivity() is used.</p>
<p>Examples.</p>
</div>
<div id="c.-identifying-enhancers-which-are-located-within-certain-window-from-tss" class="section level3">
<h3>1c. Identifying enhancers which are located within certain window from TSS</h3>
</div>
<div id="of-a-gene-of-interest" class="section level3">
<h3>of a gene of interest</h3>
<p>Many enhancers do not necessarily regulate the closest gene (Mifsud et al. 2015, Schoenfelder et al 2015, Javierre et al. 2016), but they bypass nearby genes while forming long-range interactions with targeted promoters (Tolhuis et al. 2002, Lettice et al. 2003) and vice-versa, genes are commonly regulated by more than one enhancer region (Tolhuis et al. 2002). Thus, this package allows to test all enhancers within certain range from the TSS (transcriptional start site) of a gene (by default it is +/-1Mb).</p>
<p>associateReg2Gene confuaionMatrix metaAssociations voteAssociations benchmarkGI bwToGeneExp regActivity regActivityAroundTSS voteAssociations regulatory activity quantification for pre-defined region in the genome, used sofar for predefined enhancer regions (regActivity)</p>
<p>This repo contains the code for predicting the target genes for regulatory elements</p>
<p>It contains the following:</p>
<p>regulatory activity quantification per pre-defined region (regulatoryActivity) the code base that is used to quantify enhancer activties across different markers and studies The scripts to match enhancers to potential target genes within pre-defined window around gene TSSes enhancer target prediction (targetPrediction) This should contain the code for target prediction methods performance metrics (performance): should contain the code for performance metrics.</p>
</div>
<div id="styles" class="section level2">
<h2>Styles</h2>
<p>The <code>html_vignette</code> template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:</p>
<pre><code>output: 
  rmarkdown::html_vignette:
    css: mystyles.css</code></pre>
</div>
<div id="figures" class="section level2">
<h2>Figures</h2>
<p>The figure sizes have been customised so that you can easily put two images side-by-side.</p>
<p>You can enable figure captions by <code>fig_caption: yes</code> in YAML:</p>
<pre><code>output:
  rmarkdown::html_vignette:
    fig_caption: yes</code></pre>
<p>Then you can use the chunk option <code>fig.cap = &quot;Your figure caption.&quot;</code> in <strong>knitr</strong>.</p>
</div>
<div id="more-examples" class="section level2">
<h2>More Examples</h2>
<p>You can write math expressions, e.g. <span class="math inline">\(Y = X\beta + \epsilon\)</span>, footnotes<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>, and tables, e.g. using <code>knitr::kable()</code>.</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">mpg</th>
<th align="right">cyl</th>
<th align="right">disp</th>
<th align="right">hp</th>
<th align="right">drat</th>
<th align="right">wt</th>
<th align="right">qsec</th>
<th align="right">vs</th>
<th align="right">am</th>
<th align="right">gear</th>
<th align="right">carb</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Mazda RX4</td>
<td align="right">21.0</td>
<td align="right">6</td>
<td align="right">160.0</td>
<td align="right">110</td>
<td align="right">3.90</td>
<td align="right">2.620</td>
<td align="right">16.46</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td>Mazda RX4 Wag</td>
<td align="right">21.0</td>
<td align="right">6</td>
<td align="right">160.0</td>
<td align="right">110</td>
<td align="right">3.90</td>
<td align="right">2.875</td>
<td align="right">17.02</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td>Datsun 710</td>
<td align="right">22.8</td>
<td align="right">4</td>
<td align="right">108.0</td>
<td align="right">93</td>
<td align="right">3.85</td>
<td align="right">2.320</td>
<td align="right">18.61</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>Hornet 4 Drive</td>
<td align="right">21.4</td>
<td align="right">6</td>
<td align="right">258.0</td>
<td align="right">110</td>
<td align="right">3.08</td>
<td align="right">3.215</td>
<td align="right">19.44</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>Hornet Sportabout</td>
<td align="right">18.7</td>
<td align="right">8</td>
<td align="right">360.0</td>
<td align="right">175</td>
<td align="right">3.15</td>
<td align="right">3.440</td>
<td align="right">17.02</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td>Valiant</td>
<td align="right">18.1</td>
<td align="right">6</td>
<td align="right">225.0</td>
<td align="right">105</td>
<td align="right">2.76</td>
<td align="right">3.460</td>
<td align="right">20.22</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>Duster 360</td>
<td align="right">14.3</td>
<td align="right">8</td>
<td align="right">360.0</td>
<td align="right">245</td>
<td align="right">3.21</td>
<td align="right">3.570</td>
<td align="right">15.84</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td>Merc 240D</td>
<td align="right">24.4</td>
<td align="right">4</td>
<td align="right">146.7</td>
<td align="right">62</td>
<td align="right">3.69</td>
<td align="right">3.190</td>
<td align="right">20.00</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td>Merc 230</td>
<td align="right">22.8</td>
<td align="right">4</td>
<td align="right">140.8</td>
<td align="right">95</td>
<td align="right">3.92</td>
<td align="right">3.150</td>
<td align="right">22.90</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td>Merc 280</td>
<td align="right">19.2</td>
<td align="right">6</td>
<td align="right">167.6</td>
<td align="right">123</td>
<td align="right">3.92</td>
<td align="right">3.440</td>
<td align="right">18.30</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
</tr>
</tbody>
</table>
<p>Also a quote using <code>&gt;</code>:</p>
<blockquote>
<p>“He who gives up [code] safety for [code] speed deserves neither.” (<a href="https://twitter.com/hadleywickham/status/504368538874703872">via</a>)</p>
</blockquote>
</div>
<div id="short-background-of-the-problem" class="section level2">
<h2>Short background of the problem</h2>
<div id="enhancers" class="section level3">
<h3>Enhancers</h3>
<p><strong>Enhancers</strong> represent distal regulatory regions in the genome that can be located up to 1Mb from the transcription start sites of genes (Lettice et al. 2003), increase gene expression regardless of their position, orientation and distance to the promoter (Banerji et al. 1981, Blackwood et al. 1998).</p>
</div>
<div id="enhancer-like-chromatin-marks" class="section level3">
<h3>Enhancer-like chromatin marks</h3>
<p>Different enhancer-like chromatin marks have been previously used to map these regulatory regions and assess their activity: level of histone modifications (especially <strong>H3K4me1</strong> and <strong>H3K27ac</strong>), nucleosome depletion, <strong>open chromatin accessibility</strong>, <strong>DNA methylation</strong>, nucleotide conservation, etc. (Mathelier et al., 2015). To improve reproducibility of mapping efforts, different chromatin marks were integrated in chromatin features (multiple modifications and more complex elements linked together, Stricker et al. 2017), combined with an unsupervised machine-learning approaches and higher number of cell types included in the mapping analysis (for example Ernst and Kellis 2012 used ChromHMM Core 15-state model and reported more than 900,000 potential enhancer regions across 127 epigenomes from the Roadmap Epigenomics Project (Kundaje et al. 2015).</p>
</div>
<div id="mapping-potential-gene-targets-of-an-enhancer-regions" class="section level3">
<h3>Mapping potential gene targets of an enhancer regions</h3>
<ol style="list-style-type: decimal">
<li><p>Mapping of regulatory regions and their targeted genes can be done computationally by correlating gene expression with levels of enhancer- associated chromatin features. Ernst et al. 2011 correlated gene expression with different histone modification marks, including enhancer associated marks H3K27ac and H3K4me1 (ENCODE Project Consortium 2012), Sheffield et al. 2013 correlated DNase I Hypersensitivity and gene expression, whereas Varley et al. 2013 was focused on DNA methylation. This package enables easy integration of …</p></li>
<li><p>A long-range interactions can be mediated by chromatin looping - a mechanism by which enhancers and promoters are brought together in the 3D space of a nucleus, which, at least partially, enables precise regulation of gene expression (Gorkin et al. 2014, Rennie et al. 2017). Thus information about the 3D genome architecture, generated by chromosome conformation capture and related techniques, is can be used as a proxy of enhancer-mediated regulation of gene expression or in these case <strong>to benchmark results</strong> of reg2gene modelling approach (Dekker et al. 2002, Dosie et al. 2006, Simonis et al. 2006, Fullwood et al. 2009, Lieberman-Aiden et al. 2009).</p></li>
</ol>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>A footnote here.<a href="#fnref1">↩</a></p></li>
</ol>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
